File: llm-service/integration-examples/LLMServiceClient.java
"""
Spring Boot Integration Example for RecruitPro
Java client for LLM microservice communication
"""

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.client.HttpClientErrorException;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;

import java.util.List;

@Slf4j
@Service
public class LLMServiceClient {

    @Value("${llm.service.url:http://localhost:8001}")
    private String llmServiceUrl;

    @Value("${llm.service.api-key:default-secret-key}")
    private String apiKey;

    private final RestTemplate restTemplate;

    public LLMServiceClient() {
        this.restTemplate = new RestTemplate();
    }

    // ========== Request/Response DTOs ==========

    @Data
    public static class ScheduleRequest {
        private List<String> hrAvailability;
        private List<String> candidateAvailability;
        private String interviewType;
        private Integer durationMinutes;
    }

    @Data
    public static class ScheduleResponse {
        private boolean success;
        private String selectedSlot;
        private String emailSubject;
        private String emailBody;
        private String error;
    }

    @Data
    public static class FeedbackRequest {
        private String role;
        private String interviewType;
        private String rawFeedback;
    }

    @Data
    public static class FeedbackResponse {
        private boolean success;
        private Integer technicalSkills;
        private Integer communication;
        private Integer problemSolving;
        private List<String> strengths;
        private List<String> weaknesses;
        private String finalRecommendation;
        private String confidenceLevel;
        private String error;
    }

    @Data
    public static class CandidateMessageRequest {
        private String evaluationSummary;
        private String outcome;
        private String candidateName;
    }

    @Data
    public static class MessageResponse {
        private boolean success;
        private String candidateMessage;
        private String error;
    }

    // ========== Service Methods ==========

    public ScheduleResponse scheduleInterview(ScheduleRequest request) {
        try {
            String url = llmServiceUrl + "/api/schedule-interview";
            HttpHeaders headers = createHeaders();
            HttpEntity<ScheduleRequest> entity = new HttpEntity<>(request, headers);

            ResponseEntity<ScheduleResponse> response = restTemplate.exchange(
                url,
                HttpMethod.POST,
                entity,
                ScheduleResponse.class
            );

            log.info("LLM Service: Interview scheduled successfully");
            return response.getBody();

        } catch (HttpClientErrorException e) {
            log.error("LLM Service Error: {}", e.getMessage());
            ScheduleResponse errorResponse = new ScheduleResponse();
            errorResponse.setSuccess(false);
            errorResponse.setError(e.getMessage());
            return errorResponse;
        }
    }

    public FeedbackResponse analyzeFeedback(FeedbackRequest request) {
        try {
            String url = llmServiceUrl + "/api/analyze-feedback";
            HttpHeaders headers = createHeaders();
            HttpEntity<FeedbackRequest> entity = new HttpEntity<>(request, headers);

            ResponseEntity<FeedbackResponse> response = restTemplate.exchange(
                url,
                HttpMethod.POST,
                entity,
                FeedbackResponse.class
            );

            log.info("LLM Service: Feedback analyzed successfully");
            return response.getBody();

        } catch (HttpClientErrorException e) {
            log.error("LLM Service Error: {}", e.getMessage());
            FeedbackResponse errorResponse = new FeedbackResponse();
            errorResponse.setSuccess(false);
            errorResponse.setError(e.getMessage());
            return errorResponse;
        }
    }

    public MessageResponse generateCandidateMessage(CandidateMessageRequest request) {
        try {
            String url = llmServiceUrl + "/api/candidate-message";
            HttpHeaders headers = createHeaders();
            HttpEntity<CandidateMessageRequest> entity = new HttpEntity<>(request, headers);

            ResponseEntity<MessageResponse> response = restTemplate.exchange(
                url,
                HttpMethod.POST,
                entity,
                MessageResponse.class
            );

            log.info("LLM Service: Candidate message generated successfully");
            return response.getBody();

        } catch (HttpClientErrorException e) {
            log.error("LLM Service Error: {}", e.getMessage());
            MessageResponse errorResponse = new MessageResponse();
            errorResponse.setSuccess(false);
            errorResponse.setError(e.getMessage());
            return errorResponse;
        }
    }

    public boolean healthCheck() {
        try {
            String url = llmServiceUrl + "/health";
            ResponseEntity<String> response = restTemplate.getForEntity(url, String.class);
            return response.getStatusCode() == HttpStatus.OK;
        } catch (Exception e) {
            log.error("LLM Service health check failed: {}", e.getMessage());
            return false;
        }
    }

    private HttpHeaders createHeaders() {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.set("X-API-Key", apiKey);
        return headers;
    }
}

// ========== Example Controller Usage ==========

@RestController
@RequestMapping("/api/interviews")
public class InterviewController {

    private final LLMServiceClient llmService;

    public InterviewController(LLMServiceClient llmService) {
        this.llmService = llmService;
    }

    @PostMapping("/schedule")
    public ResponseEntity<?> scheduleInterview(@RequestBody ScheduleRequest request) {
        LLMServiceClient.ScheduleResponse response = llmService.scheduleInterview(request);

        if (response.isSuccess()) {
            // Save to database, send email, etc.
            return ResponseEntity.ok(response);
        } else {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(response);
        }
    }

    @PostMapping("/analyze-feedback")
    public ResponseEntity<?> analyzeFeedback(@RequestBody FeedbackRequest request) {
        LLMServiceClient.FeedbackResponse response = llmService.analyzeFeedback(request);

        if (response.isSuccess()) {
            return ResponseEntity.ok(response);
        } else {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(response);
        }
    }
}

File: llm-service/integration-examples/llmServiceClient.ts
/**
 * TypeScript/Node.js Integration Example for RecruitPro Backend
 * This service client handles communication with the LLM microservice
 */

import axios, { AxiosInstance } from 'axios';

interface ScheduleRequest {
  hr_availability: string[];
  candidate_availability: string[];
  interview_type: string;
  duration_minutes?: number;
}

interface ScheduleResponse {
  success: boolean;
  selected_slot?: string;
  email_subject?: string;
  email_body?: string;
  error?: string;
}

interface FeedbackRequest {
  role: string;
  interview_type: string;
  raw_feedback: string;
}

interface FeedbackResponse {
  success: boolean;
  technical_skills?: number;
  communication?: number;
  problem_solving?: number;
  strengths?: string[];
  weaknesses?: string[];
  final_recommendation?: string;
  confidence_level?: string;
  error?: string;
}

interface CandidateMessageRequest {
  evaluation_summary: string;
  outcome: 'offer' | 'hold' | 'reject';
  candidate_name?: string;
}

interface MessageResponse {
  success: boolean;
  candidate_message?: string;
  error?: string;
}

export class LLMServiceClient {
  private client: AxiosInstance;

  constructor(
    baseURL: string = process.env.LLM_SERVICE_URL || 'http://localhost:8001',
    apiKey: string = process.env.LLM_API_KEY || 'default-secret-key'
  ) {
    this.client = axios.create({
      baseURL,
      headers: {
        'X-API-Key': apiKey,
        'Content-Type': 'application/json',
      },
      timeout: 30000, // 30 seconds
    });
  }

  /**
   * Schedule an interview with AI-powered slot selection
   */
  async scheduleInterview(data: ScheduleRequest): Promise<ScheduleResponse> {
    try {
      const response = await this.client.post<ScheduleResponse>(
        '/api/schedule-interview',
        data
      );
      return response.data;
    } catch (error) {
      console.error('LLM Service - Schedule Interview Error:', error);
      throw this.handleError(error);
    }
  }

  /**
   * Analyze interview feedback and generate structured evaluation
   */
  async analyzeFeedback(data: FeedbackRequest): Promise<FeedbackResponse> {
    try {
      const response = await this.client.post<FeedbackResponse>(
        '/api/analyze-feedback',
        data
      );
      return response.data;
    } catch (error) {
      console.error('LLM Service - Analyze Feedback Error:', error);
      throw this.handleError(error);
    }
  }

  /**
   * Generate candidate-facing communication message
   */
  async generateCandidateMessage(
    data: CandidateMessageRequest
  ): Promise<MessageResponse> {
    try {
      const response = await this.client.post<MessageResponse>(
        '/api/candidate-message',
        data
      );
      return response.data;
    } catch (error) {
      console.error('LLM Service - Generate Message Error:', error);
      throw this.handleError(error);
    }
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<{ status: string; llm_configured: boolean }> {
    try {
      const response = await this.client.get('/health');
      return response.data;
    } catch (error) {
      console.error('LLM Service - Health Check Error:', error);
      throw this.handleError(error);
    }
  }

  private handleError(error: any): Error {
    if (axios.isAxiosError(error)) {
      if (error.response) {
        return new Error(
          `LLM Service Error: ${error.response.status} - ${
            error.response.data?.error || error.message
          }`
        );
      } else if (error.request) {
        return new Error('LLM Service: No response received');
      }
    }
    return new Error(`LLM Service: ${error.message}`);
  }
}

// Example usage in your controllers
export async function handleInterviewScheduling(req: any, res: any) {
  const llmService = new LLMServiceClient();

  try {
    const result = await llmService.scheduleInterview({
      hr_availability: req.body.hr_availability,
      candidate_availability: req.body.candidate_availability,
      interview_type: req.body.interview_type,
      duration_minutes: req.body.duration_minutes,
    });

    if (result.success) {
      // Save to database, send email, etc.
      return res.json({
        success: true,
        data: result,
      });
    } else {
      return res.status(500).json({
        success: false,
        error: result.error,
      });
    }
  } catch (error: any) {
    return res.status(500).json({
      success: false,
      error: error.message,
    });
  }
}
